tasmota/tasmota_xdrv_driver/xdrv_16_tuyamcu_v1.ino

```diff
@@ -1262,150 +1261,173 @@ void TuyaInit(void) {
+      uint8_t dpidStart = 6;
+      while (dpidStart + 4 < Tuya.byte_counter) {
+        dpId = Tuya.buffer[dpidStart];
+        dpDataType = Tuya.buffer[dpidStart + 1];
+        uint16_t dpDataLen = Tuya.buffer[dpidStart + 2] << 8 | Tuya.buffer[dpidStart + 3];
+        const unsigned char *dpData = (unsigned char*)&Tuya.buffer[dpidStart + 4];
+        const char *dpHexData = ToHex_P(dpData, dpDataLen, hex_char, sizeof(hex_char));
 
-    if (serial_in_byte == 0x55) {            // Start TUYA Packet
-      Tuya.cmd_status = 1;
-      Tuya.buffer[Tuya.byte_counter++] = serial_in_byte;
-      Tuya.cmd_checksum += serial_in_byte;
+        if (TUYA_CMD_STATE == Tuya.buffer[3]) {
+          ResponseAppend_P(PSTR(",\"DpType%uId%u\":"), dpDataType, dpId);
+          if (TUYA_TYPE_BOOL == dpDataType && dpDataLen == 1) {
+            ResponseAppend_P(PSTR("%u"), dpData[0]);
+            DataVal = dpData[0];
+          } else if (TUYA_TYPE_VALUE == dpDataType && dpDataLen == 4) {
+            uint32_t dpValue = (uint32_t)dpData[0] << 24 | (uint32_t)dpData[1] << 16 | (uint32_t)dpData[2] << 8 | (uint32_t)dpData[3] << 0;
+            ResponseAppend_P(PSTR("%u"), dpValue);
+            DataVal = dpValue;
+          } else if (TUYA_TYPE_STRING == dpDataType) {
+            ResponseAppend_P(PSTR("\"%.*s\""), dpDataLen, dpData);
+            snprintf_P(DataStr, sizeof(DataStr), PSTR("%.*s"), dpDataLen, dpData);
+          } else if (TUYA_TYPE_ENUM == dpDataType && dpDataLen == 1) {
+            ResponseAppend_P(PSTR("%u"), dpData[0]);
+            DataVal = dpData[0];
+          } else {
+            ResponseAppend_P(PSTR("\"0x%s\""), dpHexData);
+            snprintf_P(DataStr, sizeof(DataStr), PSTR("%s"), dpHexData);
+          }
+        }
+
+        ResponseAppend_P(PSTR(",\"%d\":{\"DpId\":%d,\"DpIdType\":%d,\"DpIdData\":\"%s\""), dpId, dpId, dpDataType, dpHexData);
+        if (TUYA_TYPE_STRING == dpDataType) {
+          ResponseAppend_P(PSTR(",\"Type3Data\":\"%.*s\""), dpDataLen, dpData);
+        }
+        ResponseAppend_P(PSTR("}"));
+        dpidStart += dpDataLen + 4;
+      }
     }
-    else if (Tuya.cmd_status == 1 && serial_in_byte == 0xAA) { // Only packtes with header 0x55AA are valid
-      Tuya.cmd_status = 2;
+  }
+  ResponseAppend_P(PSTR("}}"));
 
-      Tuya.byte_counter = 0;
-      Tuya.buffer[Tuya.byte_counter++] = 0x55;
-      Tuya.buffer[Tuya.byte_counter++] = 0xAA;
-      Tuya.cmd_checksum = 0xFF;
+  if (Settings->flag3.tuya_serial_mqtt_publish) {  // SetOption66 - Enable TuyaMcuReceived messages over Mqtt
+/*
+    for (uint8_t cmdsID = 0; sizeof(TuyaExcludeCMDsFromMQTT) > cmdsID; cmdsID++){
+      if (TuyaExcludeCMDsFromMQTT[cmdsID] == Tuya.buffer[3]) {
+        isCmdToSuppress = true;
+        break;
+      }
     }
-    else if (Tuya.cmd_status == 2) {
-      if (Tuya.byte_counter == 5) { // Get length of data
-        Tuya.cmd_status = 3;
-        Tuya.data_len = serial_in_byte;
+*/
```

Comment:
Think this obsolete code fragment could be removed.

Reply:
👍